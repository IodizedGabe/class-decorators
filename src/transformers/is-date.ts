import { composePropertyDecorators, IsDateOptions } from '@hippo-oss/dto-decorators';
import { Type } from 'class-transformer';

import { createBasePropertyDecorators } from './base';

export function IsDate(options: IsDateOptions): PropertyDecorator {
    return composePropertyDecorators([
        ...createBasePropertyDecorators(options),

        // convert string to Date
        Type(() => Date),
    ]);
}
