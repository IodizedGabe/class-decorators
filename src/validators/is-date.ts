import { composePropertyDecorators, IsDateOptions } from '@hippo-oss/dto-decorators';
import { IsDate as IsDateDecorator } from 'class-validator';

import { createBasePropertyDecorators } from './base';

export function IsDate(options: IsDateOptions): PropertyDecorator {
    return composePropertyDecorators([
        ...createBasePropertyDecorators(options),

        // validate data as a Date
        IsDateDecorator({ each: !!options.isArray }),
    ]);
}
