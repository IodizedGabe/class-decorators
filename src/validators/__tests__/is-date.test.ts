import { validateSync } from 'class-validator';

import { IsDate } from '../is-date';

describe('IsDate', () => {
    it('allows optional date to be omitted', () => {
        class Fixture {
            @IsDate({
                optional: true,
            })
            property!: Date;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
    it('allows nullable date to be omitted', () => {
        class Fixture {
            @IsDate({
                nullable: true,
            })
            property!: Date;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
    it('does not allow required date to be omitted', () => {
        class Fixture {
            @IsDate({})
            property!: Date;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('does not allow non-date', () => {
        class Fixture {
            @IsDate({})
            property!: Date;
        }

        const obj = new Fixture();
        obj.property = 'foo' as unknown as Date;

        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('allows date', () => {
        class Fixture {
            @IsDate({})
            property!: Date;
        }

        const obj = new Fixture();
        obj.property = new Date('2022-03-11T00:37:19.486Z');

        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
});
