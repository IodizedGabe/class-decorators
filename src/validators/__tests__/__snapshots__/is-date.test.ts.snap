// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IsDate does not allow non-date 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDate": "property must be a Date instance",
    },
    "property": "property",
    "target": Fixture {
      "property": "foo",
    },
    "value": "foo",
  },
]
`;

exports[`IsDate does not allow required date to be omitted 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDate": "property must be a Date instance",
      "isDefined": "property should not be null or undefined",
    },
    "property": "property",
    "target": Fixture {},
    "value": undefined,
  },
]
`;
